L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,816 @@
+      hitPattern_.set(layer);
+    }
+  }
+
+  // construct TrackKFin from StubSF
+  TrackKFin::TrackKFin(const StubSF& stub, const TTTrackRef& ttTrackRef, const TTBV& maybePattern) :
+    Track(stub, ttTrackRef, maybePattern, stub.sectorPhi(), stub.sectorEta(), 0., 0., 0., 0.),
+    stubs_(setup()->numLayers()),
+    hitPattern_(0, setup()->numLayers())
+  {
+    get<3>(data_) = format(Variable::phiT, Process::mht).floating(stub.phiT());
+    get<4>(data_) = format(Variable::inv2R, Process::mht).floating(stub.inv2R());
+    get<5>(data_) = format(Variable::zT, Process::sf).floating(stub.zT());
+    get<6>(data_) = format(Variable::cot, Process::sf).floating(stub.cot());
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  // construct TrackKFin from TTTrackRef
+  TrackKFin::TrackKFin(const TTTrackRef& ttTrackRef, const DataFormats* dataFormats, const TTBV& maybePattern, double phiT, double inv2R, double zT, double cot, int sectorPhi, int sectorEta) :
+    Track(ttTrackRef, dataFormats, Process::kfin, maybePattern, sectorPhi, sectorEta, phiT, inv2R, zT, cot),
+    stubs_(setup()->numLayers()),
+    hitPattern_(0, setup()->numLayers())
+  {
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  vector<TTStubRef> TrackKFin::ttStubRefs(const TTBV& hitPattern, const vector<int>& layerMap) const {
+    vector<TTStubRef> stubs;
+    stubs.reserve(hitPattern.count());
+    for (int layer = 0; layer < setup()->numLayers(); layer++)
+      if (hitPattern[layer])
+        stubs.push_back(stubs_[layer][layerMap[layer]]->ttStubRef());
+    return stubs;
+  }
+
+  // construct TrackKF from Frame
+  TrackKF::TrackKF(const FrameTrack& frame, const DataFormats* dataFormats) :
+    Track(frame, dataFormats, Process::kf) {}
+
+  // construct TrackKF from TrackKfin
+  TrackKF::TrackKF(const TrackKFin& track, double phiT, double inv2R, double zT, double cot) :
+    Track(track, false, track.sectorPhi(), track.sectorEta(), track.phiT(), track.inv2R(), track.cot(), track.zT())
+  {
+    get<0>(data_) = abs(inv2R) < dataFormats_->format(Variable::inv2R, Process::sf).base() / 2. && abs(phiT) < dataFormats_->format(Variable::phiT, Process::sf).base() / 2.;
+    get<3>(data_) += phiT;
+    get<4>(data_) += inv2R;
+    get<5>(data_) += cot;
+    get<6>(data_) += zT;
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  // conversion to TTTrack with given stubs
+  TTTrack<Ref_Phase2TrackerDigi_> TrackKF::ttTrack(const vector<StubKF>& stubs) const {
+    TTBV hitVector(0, setup()->numLayers());
+    double chi2phi(0.);
+    double chi2z(0.);
+    vector<TTStubRef> ttStubRefs;
+    const int nLayer = stubs.size();
+    ttStubRefs.reserve(nLayer);
+    for (const StubKF& stub : stubs) {
+      hitVector.set(stub.layer());
+      const TTStubRef& ttStubRef = stub.ttStubRef();
+      chi2phi += pow(stub.phi(), 2) / setup()->v0(ttStubRef, this->inv2R());
+      chi2z += pow(stub.z(), 2) / setup()->v1(ttStubRef, this->inv2R());
```

Comment:
Should this be setup()->v1(ttStubRef, this->cot()); from the definition of v1: https://github.com/cms-L1TK/cmssw/blob/ecfaa590ef0d5c8cdad832e08ab8c9a6d4410be8/L1Trigger/TrackerDTC/src/Setup.cc#L576

Reply:
almost right, fix is pushed.